### DISCUSSION TITLE
What is a technical definition of "hydration" within NextJS?

### DISCUSSION THREAD

**Discussion by unitof** [vercel/next.js#22276] [Help] [answered]
The terms _hydration_, _rehydration_, _hydrate_, etc. appear periodically in NextJS documentation, but I'm unclear exactly what they refer to.

[For example](https://nextjs.org/docs/advanced-features/automatic-static-optimization):

> Statically generated pages are still reactive: Next.js will hydrate your application client-side to give it full interactivity.

From context clues (and ReactJS [documentation](https://reactjs.org/docs/react-dom.html#hydrate)) it seems hydration is the process of imbuing server-rendered HTML with the client-side JS necessary for dynamic elements or interaction? But I would appreciate a more technical definition. Is hydration a specific _event_? A general concept accomplished in many ways? A specific function or method? Is it just inherited from ReactJS with no changes?

Thanks dearly!

**Comment by tunesmith** [comment]
I'm trying to figure this out too. I'm honestly not sure if this is correct, so I would appreciate clarifications and corrections. I'm trying to compare this to concepts like reconciliation and First Contentful Paint.

`hydration`: This is the process of attaching react listeners to already-existing html dom nodes.

`rehydration`: This is pretty much the same thing, but is used in reference to SSR - SSR returns full html, but you still need to attach the react listeners to it. 

`Next.js-before-hydration`: this is the stage starting from the request to the server, through loading the html content in the browser to the point that the (re)-hydration process can begin. Depending on how your site is structured, this stage can end before or after the FCP is registered.

`Next.js-hydration`: See hydration above. This can end before or after FCP is registered, but I think it always ends before the onLoad timing event. 

What's left is react reconciliation - I think that if that single page loads leads to a lot of additional commits, there's a reconciliation process that happens after the end of `Next.js-hydration`.

I'm not sure how this relates to the number of commits in React Profiler for a single page load. I *think* that all commits after first commit will happen after `Next.js-hydration` begins.

**Reply by unitof** [reply]
What's "reconciliation" in your context?

**Reply by tunesmith** [reply]
When additional renders have to happen, the reconciliation process involves comparing the resultant virtual doms to see if there are actual changes and if a commit needs to happen.

**Reply by henrynnamani** [reply]
Reconciliation occurs when the virtual DOM checks the main DOM for changes so as to know what part to update

**Reply by arrickx** [reply]
thanks, and I also look into [what FCP is stands for.](https://web.dev/fcp/) The First Contentful Paint (FCP) metric measures the time from when the page starts loading to when any part of the page's content is rendered on the screen.

**Comment by DeanHristov** [comment, marked as answer]
A __hydration__ means "Each generated HTML is associated with minimal JavaScript code necessary for that page. When a page is loaded by the browser, its JavaScript code runs and makes the page fully interactive "

**Reply by larrasket** [reply]
What a bad name to give to something like that.

**Reply by Jia-A** [reply]
Oh technically they were looking at it as "We are sending dry HTML to the client and to give it some interactivity (life), we'll inject it with JS (water) and thus it's called hydrating".
But I do see your point here @larrasket.
